app/plugin_api.go

```diff
@@ -929,3 +929,125 @@ func (api *PluginAPI) PluginHTTP(request *http.Request) *http.Response {
 	api.app.ServeInterPluginRequest(responseTransfer, request, api.id, destinationPluginId)
 	return responseTransfer.GenerateResponse()
 }
+
+func (api *PluginAPI) CreateCommand(cmd *model.Command) (*model.Command, error) {
+	cmd.CreatorId = ""
+	cmd.PluginId = api.id
+
+	cmd, appErr := api.app.CreateCommand(cmd)
+
+	return cmd, normalizeAppErr(appErr)
+}
+
+func (api *PluginAPI) ListAllCommands(teamID string) ([]*model.Command, error) {
+	ret := make([]*model.Command, 0)
+
+	cmds, err := api.ListCustomCommands(teamID)
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	cmds, err = api.ListPluginCommands(teamID)
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	cmds, err = api.ListBuiltInCommands()
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	return ret, nil
+}
+
+func (api *PluginAPI) ListCustomCommands(teamID string) ([]*model.Command, error) {
+	// Plugins are allowed to bypass the a.Config().ServiceSettings.EnableCommands setting.
+	return api.app.Srv().Store.Command().GetByTeam(teamID)
+}
+
+func (api *PluginAPI) ListPluginCommands(teamID string) ([]*model.Command, error) {
+	commands := make([]*model.Command, 0)
+	seen := make(map[string]bool)
+
+	for _, cmd := range api.app.PluginCommandsForTeam(teamID) {
+		if !seen[cmd.Trigger] {
+			seen[cmd.Trigger] = true
+			commands = append(commands, cmd)
+		}
+	}
+
+	return commands, nil
+}
+
+func (api *PluginAPI) ListBuiltInCommands() ([]*model.Command, error) {
+	commands := make([]*model.Command, 0)
+	seen := make(map[string]bool)
+
+	for _, value := range commandProviders {
+		if cmd := value.GetCommand(api.app, api.app.T); cmd != nil {
+			cpy := *cmd
+			if cpy.AutoComplete && !seen[cpy.Trigger] {
+				cpy.Sanitize()
+				seen[cpy.Trigger] = true
+				commands = append(commands, &cpy)
+			}
+		}
+	}
+
+	return commands, nil
+}
+
+func (api *PluginAPI) GetCommand(commandID string) (*model.Command, error) {
+	cmd, appErr := api.app.GetCommand(commandID)
+	return cmd, normalizeAppErr(appErr)
+}
+
+func (api *PluginAPI) UpdateCommand(commandID string, updatedCmd *model.Command) (*model.Command, error) {
+	oldCmd, err := api.GetCommand(commandID)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedCmd.Trigger = strings.ToLower(updatedCmd.Trigger)
+	updatedCmd.Id = oldCmd.Id
+	updatedCmd.Token = oldCmd.Token
+	updatedCmd.CreateAt = oldCmd.CreateAt
+	updatedCmd.UpdateAt = model.GetMillis()
+	updatedCmd.DeleteAt = oldCmd.DeleteAt
+	updatedCmd.CreatorId = oldCmd.CreatorId
+	updatedCmd.PluginId = oldCmd.PluginId
```

Comment:
Should we force `api.id` here? Should we allow both `CreatorId` and `PluginId` to be set?

Reply:
I was thinking about this, and had written a DM to you, along the lines of "are we sure we want to allow a plugin to basically take ownership over any command, and actually /force/ them to take ownership if they update an existing command?", but then I realized that we basically allow them to do that anyway. They can delete a command and create a new one over it anyway. 

So I'm letting them set CreatorId, and forcing PluginId to be api.id. 

By the way, I updated the `DeleteCommand` function--it would fail if `EnableCommands` was false, so I'm bypassing that check. Same for `GetCommand` and `CreateCommand`, and then I removed NormalizeAppError (didn't need it anymore).

There's a slight problem I noticed though--some of the community work to convert to error means `appErrors` are being cast into `error`, eg: https://github.com/mattermost/mattermost-server/blob/ab2ae46c2afae5e2dda7c1e5d28de564f9020dba/store/sqlstore/command_store.go#L179

Should we do something about this? Could cause havok down the line.